### PR TITLE
Implement asynchronous support in ODataJsonLightErrorDeserializer

### PR DESCRIPTION
**Pull request by gathogojr**
### Issues

This pull request partially fulfills #2019.

### Description

Implement asynchronous support in `ODataJsonLightErrorDeserializer`.

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

**Comment by gathogojr** [comment]
Marked PR as draft until related PR is merged and rebased

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs [hunk 1]
```diff
@@ -463,5 +458,397 @@ private void ReadPropertyValueInODataErrorDetailObject(ODataErrorDetail detail,
...
+        /// Post-Condition: JsonNodeType.EndOfInput
+        /// </remarks>
+        private async Task<ODataError> ReadTopLevelErrorImplementationAsync()
+        {
+            ODataError error = null;
+
+            while (this.JsonReader.NodeType == JsonNodeType.Property)
+            {
+                string propertyName = await this.JsonReader.ReadPropertyNameAsync()
+                    .ConfigureAwait(false);
```
Comment (KenitoInc) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs:482]:
```suggestion
                    .ConfigureAwait(false);

```

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs [hunk 2]
```diff
@@ -463,5 +458,397 @@ private void ReadPropertyValueInODataErrorDetailObject(ODataErrorDetail detail,
...
+            Debug.Assert(!string.IsNullOrEmpty(propertyAnnotationName), "!string.IsNullOrEmpty(propertyAnnotationName)");
+            Debug.Assert(
+                propertyAnnotationName.StartsWith(JsonLightConstants.ODataAnnotationNamespacePrefix, StringComparison.Ordinal),
+                "The method should only be called with OData. annotations");
+            this.AssertJsonCondition(JsonNodeType.PrimitiveValue, JsonNodeType.StartObject, JsonNodeType.StartArray);
+
+            if (string.Equals(propertyAnnotationName, ODataAnnotationNames.ODataType, StringComparison.Ordinal))
+            {
+                string typeName = ReaderUtils.AddEdmPrefixOfTypeName(
+                    ReaderUtils.RemovePrefixOfTypeName(await this.JsonReader.ReadStringValueAsync().ConfigureAwait(false)));
```
Comment (KenitoInc) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs:608]:
```suggestion
                    ReaderUtils.RemovePrefixOfTypeName(await this.JsonReader.ReadStringValueAsync().ConfigureAwait(false)));

```

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs [hunk 3]
```diff
@@ -463,5 +458,397 @@ private void ReadPropertyValueInODataErrorDetailObject(ODataErrorDetail detail,
...
+                if (typeName == null)
+                {
+                    throw new ODataException(Strings.ODataJsonLightPropertyAndValueDeserializer_InvalidTypeName(propertyAnnotationName));
+                }
+
+                this.AssertJsonCondition(JsonNodeType.Property, JsonNodeType.EndObject);
```
Comment (KenitoInc) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs:614]:
```suggestion
                this.AssertJsonCondition(JsonNodeType.Property, JsonNodeType.EndObject);

```

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs [hunk 4]
```diff
@@ -463,5 +458,397 @@ private void ReadPropertyValueInODataErrorDetailObject(ODataErrorDetail detail,
...
+            this.JsonReader.AssertNotBuffering();
+
+            ValidationUtils.IncreaseAndValidateRecursionDepth(ref recursionDepth, this.MessageReaderSettings.MessageQuotas.MaxNestingDepth);
+
+            ODataInnerError innerError = new ODataInnerError();
+
+            await this.ReadJsonObjectInErrorPayloadAsync(
+                (propertyName, propertyAndAnnotationCollector) => this.ReadPropertyValueInInnerErrorAsync(recursionDepth, innerError, propertyName)).ConfigureAwait(false);
+
+            this.JsonReader.AssertNotBuffering();
```
Comment (KenitoInc) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs:662]:
```suggestion
            this.JsonReader.AssertNotBuffering();

```

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs [hunk 5]
```diff
@@ -463,5 +458,397 @@ private void ReadPropertyValueInODataErrorDetailObject(ODataErrorDetail detail,
...
+                        .ConfigureAwait(false);
+                    break;
+
+                case JsonConstants.ODataErrorDetailsName:
+                    error.Details = await this.ReadErrorDetailsAsync()
+                        .ConfigureAwait(false);
+                    break;
+
+                case JsonConstants.ODataErrorInnerErrorName:
+                    error.InnerError = await this.ReadInnerErrorAsync(0 /* recursionDepth */)
```
Comment (habbes) [inline src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs:763]:
```suggestion
                    error.InnerError = await this.ReadInnerErrorAsync(recursionDepth: 0)

```

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs (hunk 1 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs:482]: Fixed

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs (hunk 3 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs:617]: Fixed

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs (hunk 4 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs:667]: Fixed

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs (hunk 2 above)
Comment (gathogojr) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs]: Fixed

**Review by habbes — APPROVED** [review]

File: src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs (hunk 4 above)
Comment (habbes) [reply src/Microsoft.OData.Core/JsonLight/ODataJsonLightErrorDeserializer.cs:667]: What's the convention here? Should we add a new line after each statement? Or what criteria is being used to determined where to place the empty line?

**Review by KenitoInc — APPROVED** [review]

**Pull request merged**